File: bindings/pyomq/examples/zguide/03_pipeline/sink.py
"""ZGuide 03 — Sink (result collector).

PULL socket binds and collects results from workers. Prints per-worker
distribution when the expected count is reached.

    python sink.py [sink_ep] [expected_count]
"""
import sys
from collections import defaultdict

import pyomq as zmq

sink_ep = sys.argv[1] if len(sys.argv) > 1 else "ipc://@omq-zguide-03-sink"
expected = int(sys.argv[2]) if len(sys.argv) > 2 else 100

with zmq.Context() as ctx:
    pull = ctx.socket(zmq.PULL)
    pull.setsockopt(zmq.RCVTIMEO, 5000)
    pull.bind(sink_ep)

    print(f"sink: listening on {sink_ep}, expecting {expected} results")

    counts = defaultdict(int)
    received = 0

    for i in range(expected):
        try:
            body = pull.recv_string()
        except zmq.Again:
            print(f"sink: timeout after {received} results")
            break
        received += 1
        worker_id = body.split(":")[0]
        counts[worker_id] += 1
        if received % 25 == 0 or received == expected:
            print(f"sink: received {received}/{expected}")

    print(f"sink: done — {received} results from {len(counts)} workers")
    for worker, count in sorted(counts.items()):
        print(f"  {worker}: {count} items")

File: examples/zguide/03_pipeline/sink.rs
//! `ZGuide` 03 — Sink (result collector).
//!
//! PULL socket binds and collects results from workers. Prints
//! per-worker distribution when the expected count is reached.
//! The run script kills idle workers after the sink exits.
//!
//!     cargo run -p zguide-03-pipeline --bin sink [sink_ep] [expected_count]

use std::collections::HashMap;

use omq_tokio::{Context, Endpoint, Message, Options, SocketType};

fn endpoint_or(args: &[String], index: usize, default: &str) -> Endpoint {
    args.get(index).map_or_else(
        || default.parse().unwrap(),
        |s| s.parse().expect("invalid endpoint"),
    )
}

fn msg_str(msg: &Message, idx: usize) -> String {
    String::from_utf8_lossy(&msg.part_bytes(idx).unwrap()).to_string()
}

#[tokio::main(flavor = "current_thread")]
async fn main() {
    let ctx = Context::new();
    let args: Vec<String> = std::env::args().collect();
    let sink_ep = endpoint_or(&args, 1, "ipc://@omq-zguide-03-sink");
    let expected: usize = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(1000);

    let pull = ctx.socket(SocketType::Pull, Options::default());
    pull.bind(sink_ep.clone()).await.unwrap();

    println!("sink: listening on {sink_ep}, expecting {expected} results");

    let mut counts: HashMap<String, usize> = HashMap::new();

    for i in 0..expected {
        let msg = pull.recv().await.unwrap();
        let body = msg_str(&msg, 0);
        let worker_id = body.split(':').next().unwrap_or("unknown").to_string();
        *counts.entry(worker_id).or_default() += 1;
        if (i + 1) % 25 == 0 || i + 1 == expected {
            println!("sink: received {}/{expected}", i + 1);
        }
    }

    println!(
        "sink: done — {expected} results from {} workers",
        counts.len()
    );
    let mut sorted: Vec<_> = counts.into_iter().collect();
    sorted.sort();
    for (worker, count) in &sorted {
        println!("  {worker}: {count} items");
    }
}
